### PR TITLE
feat: make outline condensed + move indentation logic to CSS

### PR DESCRIPTION
**Pull request by lierdakil**
I would rather base this on #93, but I'm posting it here lest I eventually forget about this and lose the work.

TL;DR: let CSS do the indent style calculations. This might have a minor performance impact, but I can't really test for that right now.

`atom-ide-outline-item-hasChildren` class name was chosen because `hasChildren` is pretty generic, and we might run into some interference from other packages defining styles for this class.

`margin` and `padding` on `.fold` are set to 0 primarily to simplify computations. Arguably the alignment is a little more consistent this way, but that's a matter of taste I guess. The old alignment can be approximated to arbitrary precision by tweaking `margin-left` on `.fold::before`.

Waiting on #93, marking as draft.

### CODE REVIEW

**Comment by aminya** [comment]
#93 is merged. Would you like to rebase?

**Review by aminya — CHANGES_REQUESTED** [review]
When there is a fold button, the entries are not indented correctly.
image

Previously:

image

**Comment by lierdakil** [comment]
Yep, yep, I've messed the rebase up, had the stray `.fold` name used. I've amended.

There is a stylistic difference I didn't catch on immediately. Wherein before `<li>`s would go all the way to the left, now there's a gap dependent on the indent level. Which affects hover highlighting. The gap is arguably worse. Let me iterate on this for a little more.

**Comment by aminya** [comment]
The spec/test.py should look like this:

image

**Comment by aminya** [comment]
One of the other issues we have (as you see in the above image) is that the indentation becomes so much as the level increases. I tried to decrease this ratio, but then some files like json (powered by ide-json) become very squashed. If you can fix that issue too, it would be great.

**Comment by lierdakil** [comment]
Looks aligned on my end
image

But as I said, I need to iterate on this some more.

**Comment by lierdakil** [comment]
Okay, this should be okay. Could've moved counting logic to CSS too, but it needs some weird magic to work, so that's a bit too brittle for my taste.

a376740 is a bit of a request for comments. It can help with the "too much indent" issue if we add specific grammar overrides, e.g. (from the top of my head)

```less
.outline-view[data-editor-root-scope="source.python"] {
  --level-indent-size: 8px;
}
.outline-view[data-editor-root-scope="source.json"] {
  --level-indent-size: 32px;
}
```

but it's arguable whether having grammar-specific overrides is something we want to do. a376740 can be still useful for power users though.

**Comment by aminya** [comment]
The levels don't seem to work. Here is the test.py. The class properties/methods are at the same level of the class itself:

image

**Comment by lierdakil** [comment]
This was a typical off-by-one error. And I can even explain why I didn't catch that. Since I don't have a python outline provider at hand, I was testing with typescript. And typescript has only one top-level element.

Side note, for cases where there is indeed the only one top-level element, something like this could be handy:

```less
.outline-view > ul > li:only-child > ul {
  margin-left: ~"calc(0px - var(--editor-tab-length) * var(--level-indent-size))";
}
```
(or any small-ish shift instead of `0px`).

**Comment by aminya** [comment]
That's why I have added a bunch of different file formats to the spec folder 😄 I check them all to see if they all work.

**Comment by aminya** [comment]
I am going to push a commit that condenses the fold button a bit.

**Comment by aminya** [comment]
This how test.py looks like now:
image

File: styles/atom-ide-outline.less [hunk 1]
```diff
@@ -42,7 +43,7 @@
       padding: 2px 5px;
       white-space: nowrap;
 
-      padding-left: ~"calc(var(--editor-tab-length) * var(--level-indent-size) * var(--indent-level) + 5px + @{fold-width})";
+      padding-left: ~"calc(var(--editor-tab-length) * var(--level-indent-size) * var(--indent-level) + @{fold-width})";
```
Comment (lierdakil) [inline styles/atom-ide-outline.less:46]: These 5px come from `padding` definition two lines above. Since indentation is supposed to be "in addition" to any padding that already exists.

File: styles/atom-ide-outline.less [hunk 2]
```diff
@@ -16,10 +16,11 @@
 }
 
 .outline-view {
+  margin-left: 10px; // overal left margin
```
Comment (lierdakil) [inline styles/atom-ide-outline.less:19]: This creates a 10px gap between item highlighting and pane border. IMO this is ugly. I would suggest avoiding positive margins.

File: styles/atom-ide-outline.less (hunk 1 above)
Comment (lierdakil) [reply styles/atom-ide-outline.less:46]: I'm not particularly attached to those 5px, just thought I should explain where those came from.

Comment (aminya) [reply styles/atom-ide-outline.less]: I added it back in 
https://github.com/atom-community/atom-ide-outline/pull/96/commits/360249bd623c832b393a4881cf90766c0f9c8a2f

File: styles/atom-ide-outline.less (hunk 2 above)
Comment (aminya) [reply styles/atom-ide-outline.less]: It can be replaced with `padding-left: 5px;` under `.ul`

Comment (lierdakil) [reply styles/atom-ide-outline.less]: That won't help though. We need `span` items to be flush on left side, because those are the ones that are highlighted.

Gap:
image
image

Just increase `@padding-side`. Lose the margin.

No gap:
image

Comment (aminya) [reply styles/atom-ide-outline.less]: Thanks. Fixed in 
https://github.com/atom-community/atom-ide-outline/pull/96/commits/019f508f87a4e4e4b8a273cb6963436565a57cf3

**Comment by aminya** [comment]
So, to make the languages with tab-length of 4 or more take less space (without affecting those with tab-length 2), I want to use something like the following inside calc. But it doesn't seem to work. Doesn't CSS support this?
```css
max(var(--editor-tab-length)/2, 2)
```

**Review by aminya — APPROVED** [review]
This now looks great

TypeScript:
image

JSON:
image

Python:
image

**Comment by aminya** [comment]
Changed the default icon to make it look better when combined with other icons:

image

**Comment by lierdakil** [comment]
> Doesn't CSS support this?

`max` and `min` were implemented in Chrome 79. Stable Atom is still on Chrome 76. Next version reportedly uses Electron 9, so probably Atom 1.56 will support `max` (don't quote me on that)

I'm fine with leaving the js hack in until then.
